8292170: Convert CodeRootSetTable to use ResourceHashtable

lkorinth opened the pull request:
I am replacing the old hash table with ResourceHashtable. In the process I am also removing the redundant `_length` field. By removing the `_length` field, a lot of asserts can be removed as the length will trivially match the length of the underlying table.

I would like to have feedback on the addition of `unlink_destruct(Function&& should_remove)`. I added it because I prefer to use a functor object that can be used by lambdas instead of an iterator object. However, I did add code to "destruct" resource objects when they are removed instead of just letting them be to reflect that we use `delete` on c-heap objects. Maybe I should remove this "improvement" and then maybe implement the function by calling the existing unlink?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292170](https://bugs.openjdk.org/browse/JDK-8292170): Convert CodeRootSetTable to use ResourceHashtable


### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [ec36d732](https://git.openjdk.org/jdk/pull/11675/files/ec36d732d51804cbc69403f61f9cbfaca414dfd0)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11675/head:pull/11675` \
`$ git checkout pull/11675`

Update a local copy of the PR: \
`$ git checkout pull/11675` \
`$ git pull https://git.openjdk.org/jdk pull/11675/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11675`

View PR using the GUI difftool: \
`$ git pr show -t 11675`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11675.diff">https://git.openjdk.org/jdk/pull/11675.diff</a>

</details>

coleenp commented on src/hotspot/share/utilities/resourceHash.hpp:
```diff
@@ -274,6 +274,36 @@ class ResourceHashtableBase : public STORAGE {
     }
   }
 
+  // The argument should_remove() have the signature: bool Function(K const&, V const&).
```
Can you make the unlink a separate change?  I think unlink with the Function argument should be an overload of unlink() and implemented like the iterate function. I don't immediately understand the additional &&s.
If it's a separate change, we can fix the other callers to use lambdas, which would be nicer.
Also, I think calling the ResourceObj destructor is another discussion - they are not typically used and I think the discussion should be somewhere else.

lkorinth replied:
I will rework it.

coleenp reviewed (commented):
I'm happy to see this change coming!

lkorinth commented:
I changed the code to use the original iterator interface and made the closure NONCOPYABLE. Also updated copyrights.

coleenp commented on src/hotspot/share/gc/g1/g1CodeCacheRemSet.cpp:
```diff
@@ -286,11 +219,9 @@ class CleanCallback : public StackObj {
-    int removed = _table->remove_if(should_clean);
-    assert((size_t)removed <= _length, "impossible");
-    _length -= removed;
+    _table->remove_if(should_clean);
```
remove_if is a template function but it's only called here.  Does it need to be a template function?

lkorinth replied:
It does not, I will fix this.

coleenp commented on src/hotspot/share/gc/g1/g1CodeCacheRemSet.cpp:
```diff
@@ -40,104 +40,42 @@
-  }
+  _table.iterate_all([&](nmethod* nm, nmethod* _) {
+    blk->do_code_blob(nm);
+  });
```
cool use of lambdas.

coleenp reviewed (approved):
Well this looks A LOT better than the old hashtable and I'm hitting approve but I don't know this code so take that with a grain of salt.  Thank you.

walulyai commented on src/hotspot/share/gc/g1/g1CodeCacheRemSet.cpp:
```diff
@@ -40,104 +40,42 @@
 G1CodeRootSetTable* volatile G1CodeRootSetTable::_purge_list = NULL;
 
 size_t G1CodeRootSetTable::mem_size() {
```
G1CodeRootSetTable::* definitions in `g1CodeCacheRemSet.cpp` make `g1CodeRootSetTable.hpp` seem redundant. Either add `g1CodeRootSetTable.cpp` or move the `G1CodeRootSetTable` declaration to `g1CodeCacheRemSet.hpp`

lkorinth replied:
I will merge them.

walulyai commented on src/hotspot/share/gc/g1/g1CodeRootSetTable.hpp:
```diff
@@ -24,44 +24,42 @@
+  Table _table;
   G1CodeRootSetTable* _purge_next;
 
   unsigned int compute_hash(nmethod* nm) {
```
Change makes this dead code

lkorinth replied:
I do not understand, I am still "using"  Table in `mem_size`, or did you mean something else?

walulyai replied:
`unsigned int compute_hash(nmethod* nm) {` function,  sorry for not being clear

lkorinth replied:
Yes, will fix, nice find!

lkorinth commented:
Ivan has found a test error, and it looks like I have a bug.

iklam commented on src/hotspot/share/gc/g1/g1CodeCacheRemSet.cpp:
```diff
@@ -40,104 +39,41 @@
-  assert(number_of_entries() == 0, "should have removed all entries");
+  return sizeof(*this) +
+    _table.table_size() * sizeof(Table::Node*) +
+    _table.number_of_entries() * sizeof(Table::Node);
```
Instead of exporting Table::Node and the details of the implementation, maybe ResourceHashtable should have a function that returns its footprint? I don't know what name we usually use for that. Maybe `ResourceHashtable::mem_size()` to follow the example of `G1CodeRootSetTable::mem_size()`?

lkorinth replied:
I am back after vacation, I added a `ResourceHashtable::mem_size()`, does it look okay?

coleenp commented on src/hotspot/share/utilities/resourceHash.hpp:
```diff
@@ -294,6 +294,12 @@ class ResourceHashtableBase : public STORAGE {
+    return sizeof(*this) +
+      table_size() * sizeof(Node*) +
+      number_of_entries() * sizeof(Node);
+  }
```
Thank you Leo and Ioi. This was a good suggested change.

coleenp replied:
Yes, this looks right to me.

lkorinth commented:
Although I have no test problems, I am still looking into stuff that worries me. My worries are around the code cache methods that do *not* use locks (`contains` and `nmethods_do`). The earlier hash table used used `Atomic::release_store(&_entry, l)`. Interleaved (non-locked) accesses might now pick bad pointers (the reason why I removed `contains` from `add`). I will update you on my findings...

coleenp commented:
This hashtable requires locking, so your nmethods_do function must be protected by either your _m lock or maybe CodeCache_lock (?)
contains() should assert that _m lock is held.

coleenp reviewed (changes requested):
There are further cleanups to do for this.

lkorinth commented:
`code_roots_list_contains` is used in `assert`s, `gurantee`s and in heap verification code only. Make it locked so that we do not have to think about correctness as it will not affect production code (at least if we remove the `guarantee`). I also added new asserts that does fire when we do `add` to a hash table that we iterate in another thread. So we have at least one problem left that might possibly be fixed by iterating on a copy, looking into that now.

coleenp commented on src/hotspot/share/gc/g1/g1CodeRootSet.cpp:
```diff
@@ -0,0 +1,144 @@
+#include "oops/oop.inline.hpp"
+#include "runtime/atomic.hpp"
+#include "services/memTracker.hpp"
+#include "utilities/hashtable.inline.hpp"
```
Need to remove this include since it's no longer used.  Also, I don't see memTracker used either.  is stack.inline.hpp used ?

lkorinth replied:
I will cleanup `#include`s and `class`lists

coleenp commented on src/hotspot/share/gc/g1/g1CodeRootSet.hpp:
```diff
@@ -22,68 +22,59 @@
 
 #include "utilities/globalDefinitions.hpp"
 
+#include "code/codeCache.hpp"
```
is codeCache.hpp used by the header file?

lkorinth replied:
It defines `CodeBlobClosure` who's class forward declaration I removed.

coleenp commented on src/hotspot/share/gc/g1/heapRegionRemSet.cpp:
```diff
@@ -124,7 +120,10 @@ void HeapRegionRemSet::add_code_root_locked(nmethod* nm) {
           BOOL_TO_STR(_m.owned_by_self()), BOOL_TO_STR(Thread::current()->is_VM_thread()));
-  _code_roots.add(nm);
+
+  if (!_code_roots.contains(nm)) { // with this test, we can assert that we do not modify the hash table while iterating over it
```
But you have the assert in 'add' too, so put_if_absent is more efficient than calling contains first.

lkorinth replied:
When we have an evacuation failure, the code will add an nmethod that is already present. If I guard in `add_code_root_locked`, evacuation failures will not reach the assert in `G1CodeRootSet::add`. Another way to do it is to to do the `contains` within the assert and then somewhat optimising the release build. Would you prefer that?

coleenp commented on src/hotspot/share/gc/g1/g1CodeRootSet.hpp:
```diff
@@ -22,68 +22,59 @@
+#include "utilities/resizeableResourceHash.hpp"
+#include "utilities/resourceHash.hpp"
+
+class CleanCallback;
```
Do you need this forward declaration since it's in the .cpp file before the caller?

coleenp reviewed (changes requested):
This looks really good! A couple of include file cleanups needed.

coleenp reviewed (approved):
Good, thank you for the cleanups and switching to the better Hashtable.

lkorinth commented:
tier1-5 passes. @iklam and @walulyai, can you take a look at my changes? I have removed the extra table class, now doing everything in G1CodeRootSet. Therefore I also renamed the files to `g1CodeRootSet.?xx`. I have also added asserts ensuring that we do not mutate the table when we iterate over it. I also removed all logic related to the purge list as we are now doing the resizing using the resizable hash table logic. I think this is ready to be integrated.

walulyai reviewed (approved):
Lgtm!

iklam reviewed (approved):
LGTM

lkorinth commented:
Many thanks for the review work Ivan, Ioi and Coleen!!!

lkorinth commented:
/integrate
